### ISSUE TITLE
Slow down with lots of different contributions (~ @ 700 lines)

### ISSUE THREAD

**Issue by JohnMcLear** [ether/etherpad-lite#1763]
https://bugzilla.wikimedia.org/show_bug.cgi?id=46637

Please attempt to replicate and profile.

**Comment by JohnMcLear** [comment]
@dlongley Hey man just wondering if you had time to look into this further yet? :)

**Comment by dlongley** [comment]
I'm still time strapped at the moment. However, I'm thinking the right approach to solving this, probably a number of other bugs, and getting maintenance into a better state would be:

1. Rewrite the core editor using a new version of [Ace](https://github.com/ajaxorg/ace). I think etherpad-lite should have more of a dependency-based relationship with Ace (vs. what appears to be highly-modified check-ins of very old version (2009)). This would let us pull in fixes from Aces independently of etherpad-lite and off-load a ton of maintenance-related bugs, reduce etherpad-lite's code base size, keep concerns limited to etherpad-lite's value-add and therefore keep things simpler for devs, etc.
2. Etherpad-lite's core design should be around tracking, storing, and transmitting changes to an Ace document (from multiple users), nothing more. On top of the core should be the same sort of existing feature set that exists now (hopefully changing core wouldn't require us to make too many changes elsewhere). In short, after a rewrite of core, we should still have all the features and a similar feel to what we had before, but core's scope would be appropriately limited, the code would be easier to understand, and we could focus on etherpad-lite features instead of worrying about the underlying editor itself.
3. As I'm always short on time, I've only briefly looked at Ace's APIs, but it seems like the first cut at a redesign would involve writing an [EditSession](http://ace.c9.io/#nav=api&api=edit_session) that tracked changes, communicated them to the server, and received changes from other users from the server. Each user would  just get their own EditSession on a document. Changes would be serialized to/from JSON and, at least it seems, could be applied as a change set via a simple call to `applyDeltas` on the shared [Document](http://ace.c9.io/#nav=api&api=document). I haven't looked into all of what goes into an Ace change object or what etherpad-lite is currently serializing for changes; we may/may not need to update the format or write something to convert legacy pads. We'd still need a way to leave meta tags on elements that were changed when applying each delta (to colorize for different users) -- but, maybe something easy/clever can be done here as well. I didn't look into it; maybe there's even a way to abuse a custom syntax highlighter to accomplish it.

There are more details than this to work through, but this is just a starting point for considering a redesign that would free etherpad-lite from its old and seemingly-maintainable core code base. Hopefully the new core can be engineered at a much high-level of abstraction as proposed above, delegating all of the input event handling, rendering, messing with ranges and so forth to Ace. I think doing this sort of rewrite of core is probably the best approach for the long term health of etherpad-lite. And, it may turn out, that it's actually the easiest way to solve this "black hole" problem.

**Comment by JohnMcLear** [comment]
I reached out to @nightwing for 1

**Comment by JohnMcLear** [comment]
@dlongley RE 1 see http://etherpad.googlecode.com/hg/trunk/infrastructure/ace/README - the two aren't related. ACE and Ajax editor.

**Comment by dlongley** [comment]
@JohnMcLear, ah, I see that now (see also: https://groups.google.com/forum/#!topic/etherpad-lite-dev/0uU_LXp_iPg). It seemed like perhaps we were just using a very outdated version. Perhaps we should consider switching to c9's ACE editor. I'll have to do some more research into the differences; ultimately, the end goal was to push all of the editing, low-level event handling, etc. into another package and let etherpad-lite focus on real time tracking and syncing changes from multiple users in the same document. Right now, etherpad-lite requires us to get into low-level editor engineering and it would be a huge improvement if that could be avoided altogether.

So, while I haven't reviewed the differences between c9's ACE editor and (ACE - **A**ppJet **C**ode **E**ditor), I do know that c9's ACE is quite modern, actively maintained, and seems like it has an API we can take advantage of to implement etherpad-lite's features at a higher level of abstraction.

**Comment by JohnMcLear** [comment]
c9 ACE is not contentEditable, mobile support wont work.  Afaik..

**Comment by dlongley** [comment]
It seems like it is still lacking in mobile support, despite the incredible interest in this issue: https://github.com/ajaxorg/ace/issues/37

**Comment by dlongley** [comment]
Regardless, I think etherpad-lite is only going to become maintainable if we strip out any core low-level editing stuff and rely on another popular project/module for it. If mobile support is a deal breaker, then we should either see if we can help add it to ACE or choose a different editor as a dependency.

ACE has great performance, is popular (== good maintenance), and has a good API. We can check to see how other editors match up, but we must avoid projects that lack strong communities around them or that require significant custom low-level editing code to implement etherpad-lite's core features.

**Comment by tanyo520** [comment]
> 此刻，我仍然时间有限。但是，我正在考虑解决此问题的正确方法，可能是其他一些错误，并使维护处于更好的状态将是：
> 
> 1. 使用新版本的[Ace](https://github.com/ajaxorg/ace)重写核心编辑器。我认为etherpad-lite应该与Ace具有更多的基于依赖关系的关系（通过模块进行安装，而不是对一个非常老的版本（2009）进行高度修改的签入）。这样一来，我们就可以独立于Acepad-lite来从Ace中获取修复程序，并卸载大量与维护相关的错误，减少Etherpad-lite的代码库大小，将关注点限制在etherpad-lite的增值范围内，从而使事情变得更简单开发人员等
> 2. Etherpad-lite的核心设计应围绕（从多个用户）跟踪，存储和传输对Ace文档的更改，仅此而已。核心之上应该是现在存在的同类功能集（基本的UI内容，挑选颜色，查看历史记录等；希望更改核心不会要求我们在其他地方进行太多更改）。简而言之，在重写内核之后，我们仍应具有所有功能，并具有与以前相似的感觉，但是内核的范围将受到适当限制，代码将更易于理解，并且我们可以专注于etherpad-lite功能，而不用担心基础编辑器本身。
> 3. 由于我总是很忙，所以我仅简要介绍了Ace的API，但似乎重新设计的[第一步](http://ace.c9.io/#nav=api&api=edit_session)是编写一个[EditSession](http://ace.c9.io/#nav=api&api=edit_session)来跟踪更改，将更改传达给服务器，并从其他用户那里接收更改从服务器。每个用户只需在文档上获得自己的EditSession。更改将序列化到JSON或从JSON序列化，至少看起来似乎可以通过`applyDeltas`对共享[Document](http://ace.c9.io/#nav=api&api=document)的简单调用将其用作更改集。[](http://ace.c9.io/#nav=api&api=document)。我没有研究过Ace变更对象中的所有内容，也没有研究etherpad-lite当前为变更而序列化的内容。我们可能会（也可能不需要）更新格式或编写一些东西来转换旧键盘。我们仍然需要一种方法来将元标记保留在应用每个增量时已更改的元素上（以针对不同用户进行着色）-但是，也许在这里也可以轻松/巧妙地做一些事情。我没有研究它；也许甚至有一种方法可以滥用自定义语法荧光笔来完成它。
> 
> 还有更多细节需要解决，但这只是考虑重新设计的起点，该重新设计可以使etherpad-lite从其旧的，看似无法维护的核心代码库中解放出来。希望可以将新内核设计为如上所述的更高层次的抽象，将所有输入事件处理，渲染，范围混乱等委托给Ace。我认为对内核进行这种重写可能是etherpad-lite长期健康的最佳方法。而且，事实证明，这实际上是解决此“黑洞”问题的最简单方法。
> 
> 我想在某个时候自己完成此操作，但是如果有任何人可以在我之前进行此操作，请务必这样做。如果有人愿意，我也会回应评论并提供设计反馈/帮助。

**Comment by JohnMcLear** [comment]
@tanyo520 You are welcome to get started whenever you can :D

**Comment by JohnMcLear** [comment]
I spent a good few hours replicating this and created a pad which is available here: https://github.com/JohnMcLear/longpad1763

TLDR; I can't replicate in modern FF or Chrome.

Do we have good enough test coverage?  No.

**Comment by JohnMcLear** [comment]
https://github.com/ether/etherpad-lite/pull/4944 has WIP tests for this issue.
